### ISSUE TITLE
org-mode slow but fixed with garbage collection delay on mac

### ISSUE THREAD

**Issue by lionelyoung** [hlissner/doom-emacs#3108]
**Describe the feature**

I recently found a fix to my "slow org-mode issue" with a garbage collection tweak.

I had an org-mode file with ~60,000 lines, and it was unusable when bullets were collapse.  Expanding all bullets helped a little, but not much.

Here's the fix:
```
;; put off GC until 500MB of allocation or 5s of idle time, and it went from unusable to fine for me:
;; via https://www.reddit.com/r/emacs/comments/55ork0/is_emacs_251_noticeably_slower_than_245_on_windows/d8cmm7v/
(setq gc-cons-threshold (* 511 1024 1024))
(setq gc-cons-percentage 0.5)
(run-with-idle-timer 5 t #'garbage-collect)
(setq garbage-collection-messages t)

```

**System information**
<details><pre>
((emacs
  (version . "26.3")
  (features . "NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS MODULES THREADS LCMS2")
  (build . "Dec 04, 2019")
  (buildopts "--with-mac --enable-mac-app=/Users/travis/build/railwaycat/homebrew-emacsmacport/build-scripts/emacs-source/tmproot --prefix=/Users/travis/build/railwaycat/homebrew-emacsmacport/build-scripts/emacs-source/tmproot --enable-mac-self-contained --with-modules")
  (windowsys . batch)
  (daemonp))
 (doom
  (version . "2.0.9")
  (build . "HEAD -> develop, origin/develop, origin/HEAD 3a5ef316f 2020-04-19 22:04:12 -0400")
  (dir . "~/.dotfiles/doom-emacs/doom.d.symlink/"))
 (system
  (type . darwin)
  (config . "x86_64-apple-darwin18.5.0")
  (shell . "/bin/zsh")
  (uname . "Darwin 18.2.0 Darwin Kernel Version 18.2.0: Thu Dec 20 20:46:53 PST 2018; root:xnu-4903.241.1~1/RELEASE_X86_64 x86_64")
  (path "~/.pyenv/shims" "~/.pyenv/bin" "~/.poetry/bin" "~/.nvm/versions/node/v12.13.0/bin" "/usr/local/opt/php@7.1/sbin" "/usr/local/opt/php@7.1/bin" "/usr/local/opt/coreutils/libexec/gnubin" "./bin" "~/.rbenv/shims" "/usr/local/bin" "/usr/local/sbin" "~/.sfs" "~/.dotfiles/bin" "~/.bin" "~/.cabal/bin" "~/Library/Haskell/bin" "/usr/bin" "/bin" "/usr/sbin" "/sbin" "/usr/local/bin" "/Library/TeX/texbin" "/usr/local/MacGPG2/bin" "/usr/local/share/dotnet" "/opt/X11/bin" "~/.dotnet/tools" "/Library/Frameworks/Mono.framework/Versions/Current/Commands" "/Applications/Xamarin Workbooks.app/Contents/SharedSupport/path-bin" "~/.pyenv/shims" "~/.pyenv/bin" "~/.poetry/bin" "~/.nvm/versions/node/v12.13.0/bin" "/usr/local/opt/php@7.1/sbin" "/usr/local/opt/php@7.1/bin" "/usr/local/opt/coreutils/libexec/gnubin" "./bin" "~/.rbenv/shims" "/usr/local/sbin" "~/.sfs" "~/.dotfiles/bin" "~/.bin" "~/.cabal/bin" "~/Library/Haskell/bin" "/Applications/Emacs.app/Contents/MacOS/libexec" "/Applications/Emacs.app/Contents/MacOS/bin"))
 (config
  (envfile . envvar-file)
  (elc-files . 0)
  (modules :completion company ivy :ui deft doom doom-dashboard doom-quit hl-todo indent-guides modeline nav-flash neotree ophints (popup +defaults) vc-gutter vi-tilde-fringe window-select workspaces :editor (evil +everywhere) file-templates fold snippets :emacs dired electric undo vc :checkers syntax :tools (eval +overlay) lookup magit :lang data emacs-lisp ledger markdown org python sh :config (default +bindings +smartparens))
  (packages (org-brain) (blacken) (magit-gitflow) (org-roam :recipe (:host github :repo "jethrokuan/org-roam" :branch "develop")))
  (unpin "n/a")
  (elpa "org-brain" "org")))

</pre></details>

**Comment by hlissner** [comment]
I'm glad this worked for you, but I highly recommend not following this advice, generally.

TLDR: Doom users should use

```elisp
(after! gcmh
  (setq gcmh-high-cons-threshold 33554432))
```

_Do not change `gc-cons-threshold` and do not change it to such an unreasonably high number_.

# Here's why:

1. A better default for garbage collection is more nuanced than bigger value = better performance.
2. When `gc-cons-threshold` is too high, you'll get intense freezing later. If it's too low you will get intense stuttering. Doom sets it to 16mb, which appears to be the sweet spot after a ton of testing.
3. However, this "sweet spot" is not an exact science and depends on what else Emacs is doing! Try doubling it, or halving it in steps until you find a sweet spot. **Do not immediately jump to 512mb!**
4. Doom users use the [gcmh](https://gitlab.com/koral/gcmh) plugin, which tries to be clever: it raises `gc-cons-threshold` when you're using Emacs, and lowers it while it's idle. `gcmh-high-cons-threshold` is the value it uses when Emacs is used (and Doom sets this to 16mb: `16777216`). 

This is why `(run-with-idle-timer 5 t #'garbage-collect)` is unnecessary.
5. `(setq garbage-collection-messages t)` **is not an optimization, it will actually slow down the garbage collector**. Only activate it for debugging purposes!
6. `(setq gc-cons-percentage 0.5)` tends to be more harmful than helpful in long term use, and just exacerbates issue no.2. Increasing `gc-cons-threshold` (or `gcmh-high-cons-threshold`, in our case) is more than enough.

If you want better techniques for speeding up org mode, I recommend you read this question in the FAQ: https://github.com/hlissner/doom-emacs/blob/develop/docs/faq.org#why-is-scrolling-slow-in-emacsdoom

**Comment by lionelyoung** [comment]
```
(after! gcmh
  (setq gcmh-high-cons-threshold 33554432))
```

Works too.  I'll switch to this doom-approved method
